js/lib/AlphabeticIndex.js

```diff
@@ -60,159 +60,291 @@ var Collator = require("./Collator.js");
-
-	//console.log("implemented in pure JS");
-	if (!Collator.cache) {
-		Collator.cache = {};
-	}
-
+	
 	Utils.loadData({
-		object: Collator, 
-		locale: this.locale, 
+		object: "AlphabeticIndex",
+		locale: this.locale,
 		name: "collation.json",
-		sync: sync,
-		loadParams: loadParams, 
+		sync: this.sync,
+		loadParams: this.loadParams, 
+
 		callback: ilib.bind(this, function (collation) {
 			if (!collation) {
 				collation = ilib.data.collation;
 				var spec = this.locale.getSpec().replace(/-/g, '_');
-				Collator.cache[spec] = collation;
+				ilib.data.cache.Collator[spec] = collation;
 			}
+			this.collation = collation;
 			this._init(collation);
-			new LocaleInfo(this.locale, {
-				sync: sync,
-				loadParams: loadParams,
-				onLoad: ilib.bind(this, function(li) {
-					this.li = li;
-					if (this.ignorePunctuation) {
-		    			isPunct._init(sync, loadParams, ilib.bind(this, function() {
-							if (options && typeof(options.onLoad) === 'function') {
-								options.onLoad(this);
-							}
-		    			}));
-	    			} else {
-						if (options && typeof(options.onLoad) === 'function') {
-							options.onLoad(this);
-						}
-	    			}
-	    		})
-			});
+
+			if (options && typeof(options.onLoad) === 'function') {
+				options.onLoad(this);
+			}
 		})
 	});
 };
 
+/**
+ * @private
+ */
+AlphabeticIndex.prototype._updateCollationMap = function(style) {
+	this.collationMap = this.collation[style].map;
+	this.flowBoundaries = this.collation[style].flowBoundaries;
+	this.indexUnits = this.collation[style].indexUnits;
+	if (this.collation[style].inherit) {
+		this.inherit = this.collation[style].inherit;
+	}
+}
+
+/**
+ * @private
+ */
+AlphabeticIndex.prototype._init = function(collation) {
+	if (!collation) {
+		return;
+	}
+
+	this.flowBoundaries = new Array();
+	
+	if (this.style === 'standard') {
+		this.style = this.collation["default"];
+	}
+	
+	this.collationMap = collation[this.style].map;
+	this.flowBoundaries = collation[this.style].flowBoundaries;
+	this.indexUnits = collation[this.style].indexUnits;
+	if (collation[this.style].inherit) {
+		this.inherit = collation[this.style].inherit;	
+	}
+	
+	if (this.collationMap === undefined &&
+		typeof(collation[this.style]) === 'string') {
+		this._updateCollationMap(collation[this.style]);
+	}
+}
+
+/**
+ * @private
+ */
+AlphabeticIndex.prototype._getKeyByValue = function(value) {
+	var i,label;
+
+	if (!value || (!ilib.isArray(value))) {
+		return "";
+	}
+
+	if (this.inheritStyle) {
+		this.collationMap = this.collation[this.inheritStyle].map;
+	}
+
+	for (i=0; i < this.indexUnits.length; i++) {
+		if (this.collationMap[this.indexUnits[i]][0]
+			=== value[0]) {
+			label = this.indexUnits[i];
+			break;
+		}
+	}
+
+	return label;
+};
+
+/**
+ * @private
+ */
+AlphabeticIndex.prototype._normalizeHangul = function(value) {
+	if (!value) {
+		return "";
+	}
+	var source = new NormString(value);
+	var normString = source.normalize("nfkd");
```

Comment:
Can we use NormString._decomposeHangul() directly instead of normalizing the whole string?

Reply:
It's not normalizing the first string not the whole string. so I think it's ok. no?